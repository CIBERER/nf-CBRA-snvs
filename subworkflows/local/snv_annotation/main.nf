include { FORMAT2INFO                                                }      from '../../../modules/local/format2info/main'
include { AUTOMAP                                                }      from '../../../modules/local/automap/main'
include { TABIX_TABIX } from '../../../modules/nf-core/tabix/tabix/main'
include { ENSEMBLVEP_VEP                                                }      from '../../../modules/nf-core/ensemblvep/vep/main'
include { POSTVEP } from '../../../modules/local/postvep/main'

workflow SNV_ANNOTATION {

    take:
    ch_vcf        // channel (mandatory): [ val(meta), path(vcf), path(tbi) ]
    ch_fasta      // channel (mandatory): [ val(meta), path(fasta) ]
    genome       // channel (mandatory): [ val(genome) ]
    species     // channel (mandatory): [ val(species) ]
    cache_version                      // channel (mandatory): [ val(cache_version) ]
    ch_cache_path                            // channel (mandatory): [ path(cache_path) ]
    ch_custom_extra_files            // channel (optional)  : [ val(meta), path(custom_extra_files) ]
    ch_extra_files                   // channel (optional)  : [ path(extra_files) ]  
    maf

    main:

    ch_versions = Channel.empty()
    //ch_vcf.view()
    //ch_extra_files.view()

    ucsc_genome = genome.map { genome_val ->
        switch(genome_val) {
            case 'GRCh38':
                return 'hg38'
            case 'GRCh37':
                return 'hg19'
            default:
                error "Unsupported genome version: ${genome_val}"
        }
    }

    FORMAT2INFO (
         ch_vcf
    )
    ch_versions = ch_versions.mix(FORMAT2INFO.out.versions.first())

    AUTOMAP (
        ch_vcf.map { meta, vcf, tbi -> [meta, vcf] },
        ucsc_genome
    )

    TABIX_TABIX(
        FORMAT2INFO.out.vcf_to_annotate
    )
    ch_versions = ch_versions.mix(TABIX_TABIX.out.versions.first())


    // ###################
    // ESTOY INTENTANDO QUE FUNCIONE DESCARGANDO EL CACHE, PERO VA A TARDAR MUCHO, ASÍ QUE PROBAR A PONERLE EL PATH THE VEP.CACHE - '/mnt/genetica5/vep_2024/ 
    // ASSEMBLY hg38
    // genome = 'GRCh38'
    // specie = "homo_sapiens" 

    //cache.view()

    //ch_custom_extra_files.view()
    //ch_custom_extra_files_and_info = ch_custom_extra_files.join(FORMAT2INFO.out.vcf_to_annotate).join(FORMAT2INFO.out.fields).join(TABIX_TABIX.out.tbi)//.view()
    ch_custom_extra_files_and_info = FORMAT2INFO.out.vcf_to_annotate.join(FORMAT2INFO.out.fields).join(TABIX_TABIX.out.tbi).join(ch_custom_extra_files).
        map { tuple ->
        def meta = tuple[0]
        def files = tuple[1..-1].flatten()
        [meta] + files
    }//.view()

    //ch_vcf.map { meta, vcf, tbi -> [meta , vcf] }//.view()
    ch_vep = ch_vcf.map { meta, vcf, tbi -> [meta , vcf] }.join(ch_custom_extra_files_and_info).map { items ->
        def meta = items[0]
        def file1 = items[1]
        def restFiles = items[2..-1]  // All files from index 2 to the end
        [
            meta,
            [file1],
            restFiles
        ]
    }//.view()
    
    ENSEMBLVEP_VEP (
        ch_vep,
        genome,
        species,
        cache_version,
        ch_cache_path,
        ch_fasta,
        ch_extra_files
    )
    ch_versions = ch_versions.mix(ENSEMBLVEP_VEP.out.versions.first())
    
    ENSEMBLVEP_VEP.out.tab.view()

    // Create a complete channel with all metadata from VEP output
    complete_ch = ENSEMBLVEP_VEP.out.tab
        .join(AUTOMAP.out.roh_automap_file, remainder: true)
        .map { meta, vep_file, automap_file ->
            // Replace null automap_file with empty list
            automap_file = automap_file ?: []
            [meta, vep_file, automap_file]
        }
    
    //complete_ch.view()

    POSTVEP (
        complete_ch,
        maf, 
        ucsc_genome
    )

    
    tsv = POSTVEP.out.pvm_tsv
    tsv.view()

    emit:
    tsv // channel: [ val(meta), path(vcf), path(tbi)]
    versions = ch_versions                     // channel: [ versions.yml ]

}