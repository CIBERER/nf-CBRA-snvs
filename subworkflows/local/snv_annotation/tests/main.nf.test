nextflow_workflow {

    name "Test Subworkflow SNV_ANNOTATION"
    script "../main.nf"
    config "./nextflow.config"

    workflow "SNV_ANNOTATION"
    tag "subworkflows"
    tag "snv_annotation"


    setup {
        
        run("ENSEMBLVEP_DOWNLOAD") {
            script "../../../../modules/nf-core/ensemblvep/download/main.nf"

            process {
                """
                input[0] = Channel.of([
                    [id:"homo_sapiens_hg38"],
                    "GRCh38",
                    "homo_sapiens_refseq",
                    "113"
                ])
                """
            }
        }

        run("TABIX_BGZIP") {
            script "../../../../modules/nf-core/tabix/bgzip/main.nf"

            process {
                """
                input[0] = Channel.value([
                    [id:"fasta"],
                    file("https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr1.fa.gz")])
                """
            }
        }
    } 

    test("annotation download cache - vcf") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test'], // meta map
                    file("https://zenodo.org/records/15730236/files/NA12878-HG001.hg38.final.chr1.chr22.vcf.gz?download=1", checkIfExists: true),
                    file("https://zenodo.org/records/15730236/files/NA12878-HG001.hg38.final.chr1.chr22.vcf.gz.tbi?download=1.tbi", checkIfExists: true)
                ])
                input[1] = TABIX_BGZIP.out.output
                input[2] = Channel.value('GRCh38')
                input[3] = "homo_sapiens"
                input[4] = "113"
                input[5] = ENSEMBLVEP_DOWNLOAD.out.cache.map{ meta, cache -> [cache] }
                input[6] = Channel.of([ id: 'test' ], [])
                input[7] = []
                input[8] = 0.01
                input[9] = file("https://raw.githubusercontent.com/TBLabFJD/GLOWgenes/master/precomputed_panelAPP/GLOWgenes_prioritization_Acute_rhabdomyolysis_GA.txt", checkIfExists: true) 
                input[10] = file("https://raw.githubusercontent.com/TBLabFJD/GLOWgenes/refs/heads/master/SGDS.csv", checkIfExists: true) 
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }


    test("annotation local cache - vcf") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test'], // meta map
                    file("https://zenodo.org/records/15730236/files/NA12878-HG001.hg38.final.chr1.chr22.vcf.gz?download=1", checkIfExists: true),
                    file("https://zenodo.org/records/15730236/files/NA12878-HG001.hg38.final.chr1.chr22.vcf.gz.tbi?download=1.tbi", checkIfExists: true)
                ])
                input[1] = TABIX_BGZIP.out.output
                input[2] = Channel.value('GRCh38')
                input[3] = "homo_sapiens"
                input[4] = "112"
                input[5] = file("/mnt/genetica5/vep_2024")
                input[6] = Channel.of([ id: 'test' ], [])
                input[7] = []
                input[8] = 0.01
                input[9] = file("https://raw.githubusercontent.com/TBLabFJD/GLOWgenes/master/precomputed_panelAPP/GLOWgenes_prioritization_Acute_rhabdomyolysis_GA.txt", checkIfExists: true) 
                input[10] = file("https://raw.githubusercontent.com/TBLabFJD/GLOWgenes/refs/heads/master/SGDS.csv", checkIfExists: true) 
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

}