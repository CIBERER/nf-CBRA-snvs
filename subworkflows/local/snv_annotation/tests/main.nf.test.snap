{
    "annotation local cache - vcf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.hg38.SNV.INDEL.annotated.tsv:md5,4944b0d79e97b5b31303ab7b4c55c38e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1a72c9e85c065be779a12a8fe91732d8",
                    "versions.yml:md5,867ea1c56daa4de47bdd9cd23bb942d9",
                    "versions.yml:md5,ef4a7940a9596752ce3a0a284652ecfa"
                ],
                "tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test.hg38.SNV.INDEL.annotated.tsv:md5,4944b0d79e97b5b31303ab7b4c55c38e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1a72c9e85c065be779a12a8fe91732d8",
                    "versions.yml:md5,867ea1c56daa4de47bdd9cd23bb942d9",
                    "versions.yml:md5,ef4a7940a9596752ce3a0a284652ecfa"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-07-03T14:00:55.613751099"
    },
    "annotation download cache - vcf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.hg38.SNV.INDEL.annotated.tsv:md5,4944b0d79e97b5b31303ab7b4c55c38e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1a72c9e85c065be779a12a8fe91732d8",
                    "versions.yml:md5,867ea1c56daa4de47bdd9cd23bb942d9",
                    "versions.yml:md5,ef4a7940a9596752ce3a0a284652ecfa"
                ],
                "tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test.hg38.SNV.INDEL.annotated.tsv:md5,4944b0d79e97b5b31303ab7b4c55c38e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1a72c9e85c065be779a12a8fe91732d8",
                    "versions.yml:md5,867ea1c56daa4de47bdd9cd23bb942d9",
                    "versions.yml:md5,ef4a7940a9596752ce3a0a284652ecfa"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-07-03T13:22:52.303051875"
    }
}